### PR TITLE
nydusify: support hook plugin

### PR DESCRIPTION
**Pull request by imeoer**
Nydusify supports the hook function execution as [go-plugin](https://github.com/hashicorp/go-plugin)
at key stages of image conversion.

The `BeforePushManifest` and `AfterPushManifest` hooks are supported
first, they will be called before/after the manifest of nydus image is
pushed to remote.

A possible use case using these hooks:

Tell the blob garbage collection service to record nydus
image reference and blob list before the image is pushed
to remote registry.

### CODE REVIEW

File: contrib/nydusify/Makefile [hunk 1]
```diff
@@ -13,10 +13,12 @@ endif
 all: build
 
 build:
-	@CGO_ENABLED=0 ${PROXY} GOOS=linux GOARCH=${GOARCH} go build -ldflags '-X main.versionGitCommit=${GIT_COMMIT} -X main.versionBuildTime=${BUILD_TIME}' -gcflags=all="-N -l" -o ./cmd ./cmd/nydusify.go
+	# Enable CGO to support go plugin
+	@CGO_ENABLED=1 ${PROXY} GOOS=linux GOARCH=${GOARCH} go build -ldflags '-X main.versionGitCommit=${GIT_COMMIT} -X main.versionBuildTime=${BUILD_TIME}' -gcflags=all="-N -l" -o ./cmd ./cmd/nydusify.go
```
Comment (changweige) [inline contrib/nydusify/Makefile:17]: Then we can't have static nydusify anymore

File: contrib/nydusify/Makefile [hunk 2]
```diff
@@ -13,10 +13,12 @@ endif
...
 
 release:
-	@CGO_ENABLED=0 ${PROXY} GOOS=linux GOARCH=${GOARCH} go build -ldflags '-X main.versionGitCommit=${GIT_COMMIT} -X main.versionBuildTime=${BUILD_TIME} -s -w -extldflags "-static"' -o ./cmd ./cmd/nydusify.go
+	# Enable CGO to support go plugin
+	@CGO_ENABLED=1 ${PROXY} GOOS=linux GOARCH=${GOARCH} go build -ldflags '-X main.versionGitCommit=${GIT_COMMIT} -X main.versionBuildTime=${BUILD_TIME} -s -w -extldflags "-static"' -o ./cmd ./cmd/nydusify.go
```
Comment (changweige) [inline contrib/nydusify/Makefile:21]: I tried to build via `make release` but errors come

```
# Enable CGO to support go plugin
# command-line-arguments
/usr/bin/ld: /tmp/go-link-4092473211/000006.o: in function `pluginOpen':
/_/plugin/plugin_dlopen.go:19: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: /tmp/go-link-4092473211/000023.o: in function `mygetgrouplist':
/_/os/user/getgrouplist_unix.go:15: warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: /tmp/go-link-4092473211/000022.o: in function `mygetgrgid_r':
/_/os/user/cgo_lookup_unix.go:37: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: /tmp/go-link-4092473211/000022.o: in function `mygetgrnam_r':
/_/os/user/cgo_lookup_unix.go:42: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: /tmp/go-link-4092473211/000022.o: in function `mygetpwnam_r':
/_/os/user/cgo_lookup_unix.go:32: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: /tmp/go-link-4092473211/000022.o: in function `mygetpwuid_r':
/_/os/user/cgo_lookup_unix.go:27: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: /tmp/go-link-4092473211/000004.o: in function `_cgo_2ac87069779a_C2func_getaddrinfo':
/tmp/go-build/cgo-gcc-prolog:58: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

Comment (changweige) [reply contrib/nydusify/Makefile:21]: Moreover, better not to print comments in makefile

File: contrib/nydusify/pkg/converter/converter.go [hunk 3]
```diff
@@ -343,10 +343,20 @@ func (cvt *Converter) convert(ctx context.Context) (retErr error) {
 		})
 	}
 
+	// Ensure all blobs can be uploaded to storage backend.
 	if err := cvt.storageBackend.Finalize(false); err != nil {
 		return errors.Wrap(err, "Finalize backend upload")
 	}
 
+	// Call hook function before push manifest to registry.
```
Comment (changweige) [inline contrib/nydusify/pkg/converter/converter.go:351]: before pushing

File: contrib/nydusify/Makefile (hunk 1 above)
Comment (imeoer) [reply contrib/nydusify/Makefile]: Fixed, using `https://github.com/hashicorp/go-plugin/`.

File: contrib/nydusify/Makefile (hunk 2 above)
Comment (imeoer) [reply contrib/nydusify/Makefile]: Fixed.

File: contrib/nydusify/pkg/converter/converter.go (hunk 3 above)
Comment (imeoer) [reply contrib/nydusify/pkg/converter/converter.go]: Fixed.

File: contrib/nydusify/pkg/hook/hook.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	logrus.Infof("[HOOK] Loaded hook plugin %s", hookPluginPath)
+
+	Caller = raw.(Hook)
+}
+
+func Close() {
+	if client != nil {
+		defer client.Kill()
```
Comment (changweige) [inline contrib/nydusify/pkg/hook/hook.go:156]: It looks so strange... Why not directly call `client.Kill()` ?

File: contrib/nydusify/pkg/hook/hook.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+func Init() {
+	if Caller != nil {
+		return
+	}
+
+	if _, err := os.Stat(hookPluginPath); err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return
+		}
+		logrus.Errorln(errors.Wrapf(err, "try load hook plugin %s", hookPluginPath))
```
Comment (changweige) [inline contrib/nydusify/pkg/hook/hook.go:118]: try loading

File: contrib/nydusify/pkg/converter/converter.go [hunk 6]
```diff
@@ -392,6 +408,9 @@ func (cvt *Converter) convert(ctx context.Context) (retErr error) {
 
 // Convert converts source image to target (Nydus) image
 func (cvt *Converter) Convert(ctx context.Context) error {
+	hook.Init()
```
Comment (changweige) [inline contrib/nydusify/pkg/converter/converter.go:411]: `Init()` will be called automatically when importing hook package?

Comment (changweige) [reply contrib/nydusify/pkg/converter/converter.go:411]: oh, it should be `init()` ignore this

**Review by changweige — APPROVED** [review]

File: contrib/nydusify/pkg/hook/hook.go (hunk 4 above)
Comment (imeoer) [reply contrib/nydusify/pkg/hook/hook.go:156]: The client may not be initialized.

**Pull request merged**